diff --git a/src/task_list/task_list.hpp b/src/task_list/task_list.hpp
index d71ae3ea..931b17c2 100644
--- a/src/task_list/task_list.hpp
+++ b/src/task_list/task_list.hpp
@@ -146,6 +146,7 @@ class TimeIntegratorTaskList : public TaskList {
 
   // functions
   TaskStatus ClearAllBoundary(MeshBlock *pmb, int stage);
+  TaskStatus RestartReceiving(MeshBlock *pmb, int stage);
 
   TaskStatus CalculateHydroFlux(MeshBlock *pmb, int stage);
   TaskStatus CalculateEMF(MeshBlock *pmb, int stage);
@@ -387,5 +388,15 @@ const TaskID RECV_RADSH(73);
 
 const TaskID SRCTERM_IMRAD(74);
 
+const TaskID SEND_HYD2(101);
+const TaskID RECV_HYD2(102);
+const TaskID SETB_HYD2(103);
+const TaskID CLEAR_ALLBND2(104);
+const TaskID RESTART_RECV(105);
+
+const TaskID SEND_SCLR2(106);
+const TaskID RECV_SCLR2(107);
+const TaskID SETB_SCLR2(108);
+
 }  // namespace HydroIntegratorTaskNames
 #endif  // TASK_LIST_TASK_LIST_HPP_
